### PR TITLE
trace: Add Kconfig options to enable verbose tracing only from certain classes

### PR DESCRIPTION
**Pull request by akloniex**
Currently, after enabling TRACEV via Kconfig, message log is flooded
with various messages, most of which do not help to debug specific
issues, due to vast volume of messages sent. This patch adds Kconfig
options for each TRACE_CLASS, to enable logs from this class specifically.

Signed-off-by: ArturX Kloniecki <arturx.kloniecki@linux.intel.com>

As next step in trace verbosity overhaul, I would like to take advantage of this feature, and degrade many of trace messages (e.g. reporting function entry) to Verbose level, which would reduce noise in production logs, lower used bandwidth, while still providing developers the possibility to cleanly enable certain log classes for debugging purposes.

### CODE REVIEW

File: src/audio/dai.c [hunk 1]
```diff
@@ -32,20 +32,17 @@
 #define trace_dai(__e, ...)					\
 	trace_event(TRACE_CLASS_DAI, __e, ##__VA_ARGS__)
 #define trace_dai_with_ids(comp_ptr, __e, ...)			\
-	trace_event_comp(TRACE_CLASS_DAI, comp_ptr,		\
-			 __e, ##__VA_ARGS__)
+	trace_event_comp(DAI, comp_ptr, __e, ##__VA_ARGS__)
```
Comment (paulstelian97) [inline src/audio/dai.c]: Wait, so this "DAI" is not defined but due to how it is used it's actually concatenated to TRACE_CLASS_DAI later on? I find that slightly confusing/sneaky.

File: src/include/sof/trace/trace-verbose-private.h [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+ */
+
+#ifndef __SOF_TRACE_TRACE_VERBOSE_PRIVATE_H__
+#define __SOF_TRACE_TRACE_VERBOSE_PRIVATE_H__
+
+#include <config.h>
+
+#if CONFIG_TRACE_VERBOSE_IRQ
+#define tracev_TRACE_CLASS_IRQ(...)		\
+	trace_event(__VA_ARGS__)
```
Comment (paulstelian97) [inline src/include/sof/trace/trace-verbose-private.h]: Wait, so how does trace_event still get the class? It's required for `sof-logger` tool (it displays the class for each event).

File: src/include/sof/trace/trace.h [hunk 3]
```diff
@@ -287,7 +288,7 @@ do {									\
 		uint32_t line_idx;				\
 		uint32_t file_name_len;				\
 		uint32_t text_len;				\
-		const char file_name[sizeof(RELATIVE_FILE)];		\
+		const char file_name[sizeof(RELATIVE_FILE)];	\
```
Comment (paulstelian97) [inline src/include/sof/trace/trace.h]: Pass-by coding style corrections should be separate commits I think?

File: src/include/sof/trace/trace.h [hunk 4]
```diff
@@ -318,11 +319,14 @@ do {									\
 
 /* verbose tracing */
 #if CONFIG_TRACEV
-#define tracev_event(...) trace_event(__VA_ARGS__)
-#define tracev_event_with_ids(...) trace_event_with_ids(__VA_ARGS__)
-#define tracev_event_atomic(...) trace_event_atomic(__VA_ARGS__)
-#define tracev_event_atomic_with_ids(...) \
-	trace_event_atomic_with_ids(__VA_ARGS__)
+#define tracev_event(class, ...)		\
+	tracev_##class(class, __VA_ARGS__)
```
Comment (paulstelian97) [inline src/include/sof/trace/trace.h]: Oh, so you do still pass the class in here.

File: src/trace/Kconfig [hunk 5]
```diff
@@ -31,4 +24,249 @@ config TRACEM
...
 	  Sending all traces by mailbox additionally.
 
+config TRACEV
+	bool "Trace verbose"
+	depends on TRACE
+	default n
+	help
+	  Enabling verbose traces.
+
+config TRACE_VERBOSE_IRQ
```
Comment (paulstelian97) [inline src/trace/Kconfig]: Submenu?

File: src/audio/dai.c (hunk 1 above)
Comment (paulstelian97) [reply src/audio/dai.c]: But I do see a reasoning as well, to not have it automatically expand to the trace class number itself. I wonder if there is a way to refactor this in another way though.

File: src/trace/Kconfig (hunk 5 above)
Comment (akloniex) [reply src/trace/Kconfig]: @paulstelian97 
I personally don't feel the need for submenu, especially, since these _depend_ on TRACEV, they don't clutter the menu, until user decides to enable TRACEV, at which point these options appear instantly, allowing user to pick trace classes to enable.

File: src/audio/dai.c (hunk 1 above)
Comment (akloniex) [reply src/audio/dai.c]: @paulstelian97 It would have worked properly, if it wasn't for `trace_event_comp` proxy macro, causing premature expansion, as you've noticed. I could've made it work with `META_DEFER` macros, iff trace classes were defined with _()_. But I figured that changing trace class definitions would cause much more hassle, than a little quirky concatenation under the hood.

File: src/trace/Kconfig (hunk 5 above)
Comment (paulstelian97) [reply src/trace/Kconfig]: Fair enough then.

File: src/audio/dai.c (hunk 1 above)
Comment (paulstelian97) [reply src/audio/dai.c]: I see. Then I don't see anything that can be easily solved so I'll approve now.

**Comment by lgirdwood** [comment]
@akloniex can you report on CI

**Comment by lgirdwood** [comment]
SOFCI TEST

File: src/trace/Kconfig (hunk 5 above)
Comment (jajanusz) [reply src/trace/Kconfig]: @akloniex Can you surround them with `if TRACEV` & `endif # TRACEV` ?

Comment (jajanusz) [reply src/trace/Kconfig]: If applies `depends` and there is no need to repeat it on every entry

File: src/include/sof/trace/trace-verbose-private.h [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+ *
+ * Author: Artur Kloniecki <arturx.kloniecki@linux.intel.com>
+ */
+
+#ifndef __SOF_TRACE_TRACE_VERBOSE_PRIVATE_H__
+#define __SOF_TRACE_TRACE_VERBOSE_PRIVATE_H__
+
+#include <config.h>
+
+#define unused(...) do { UNUSED(__VA_ARGS__); } while (0)
```
Comment (paulstelian97) [inline src/include/sof/trace/trace-verbose-private.h]: I feel this one can be moved to a public header or UNUSED itself can be updated.

File: src/trace/Kconfig (hunk 5 above)
Comment (akloniex) [reply src/trace/Kconfig]: done

Comment (akloniex) [reply src/trace/Kconfig]: done

File: src/audio/asrc/asrc.c [hunk 7]
```diff
@@ -29,20 +29,17 @@
 #define trace_asrc(__e, ...)					\
 	trace_event(TRACE_CLASS_SRC, __e, ##__VA_ARGS__)
 #define trace_asrc_with_ids(comp_ptr, __e, ...)			\
-	trace_event_comp(TRACE_CLASS_SRC, comp_ptr,		\
-			 __e, ##__VA_ARGS__)
+	trace_event_comp(SRC, comp_ptr, __e, ##__VA_ARGS__)
```
Comment (lyakh) [inline src/audio/asrc/asrc.c]: Sorry, but I'm actually not in favour of this change (and others like this throughout the patch. The reasons are very simple: (1) it makes it more difficult to grep for these symbols, e.g. if you want to find all uses of `TRACE_CLASS_SRC` you won't find this any more. Also it makes following symbols in editors more difficult. (2) it makes it look different from the surrounding lines, e.g. the line above, which uses `TRACE_CLASS_SRC` in full.
If (1) is a trade-off - you win code compactness but you lose clarity, I think (2) is really not good - either convert all of them, or leave them all as is, don't mix them.

Comment (paulstelian97) [reply src/audio/asrc/asrc.c]: Using "TRACE_CLASS_SRC" directly will not work correctly as that will be expanded to a number and will break the very idea of this PR, to disable verbose traces for particular classes. Maybe TRACE_CLASS_SRC with some prefix or suffix?

Comment (akloniex) [reply src/audio/asrc/asrc.c]: @lyakh It was already picked up by @paulstelian97  and explained on other occurence of this change, here: https://github.com/thesofproject/sof/pull/2133#discussion_r349872229

The intermediate macro `trace_event_comp`, causes premature `TRACE_CLASS_*` token expansion, which could be solved by quite some dirty work under the hood in `trace_event` macro family expansions, or (in my opinion) more elegantly, here.
One could argue that greping for `TRACE_CLASS_*` doesn't make much sense anymore, as we've already wrapped (almost, reaching 100% soon) all traces in their respective `trace_*` macros.

If you have an idea how to solve token pasting and expansion more elegantly, I'm open for suggestions, but I honestly think that this route is cleaner than appending `_` to tokens to prevent their expansion, and then additionally accounting for it in `trace-verbose-private.h`.

Comment (lyakh) [reply src/audio/asrc/asrc.c]: @paulstelian97 thanks for the explanation, I see now. But now my question is then: is it really worth it? Is it really worth adding all those levels of macros if we anyway have the trace-verbose-private.h header, which has a bunch of defines for each of the `CONFIG_TRACE_VERBOSE_*` options? Wouldn't it be much easier and compact to just add
```
#if CONFIG_TRACE_VERBOSE_xxx
#define trace_asrc_with_ids(comp_ptr, __e, ...) ...
#else
#define trace_asrc_with_ids(...) do {} while (0)
#endif
```
in every affected file? Most of them would then just involve changing a couple of lines at the top, others, calling generic macros directly from the code, like dai.c, without wrapping them, will need such a define added at the top, which would (arguably) also simplify them?

Comment (paulstelian97) [reply src/audio/asrc/asrc.c]: @lyakh Tbh if we're doing that I'd say having the definition for `trace_asrc_with_ids` and every other such macro in a common header and make those macros public would be an easy way out. That sure is a large redo of this change and everyone would thus have to use these custom functions instead of the raw trace_event functions to benefit from that disabling.

Also in case we go with that, use the UNUSED macro instead of a do-while-0 loop. It was built specifically for using in disabled traces.

**Comment by plbossart** [comment]
@akloniex do I get this right that when the traces are disabled they are compiled out?

If yes, I am worried about Heisenbugs due to the memory layout change and timing changes. it'd be more interesting to jump to the trace and do nothing, which would also help move to dynamic debug at some point where the decision is based on a variable set at runtime instead of a constant set at build time. It's really a difference between removing traces and filtering their output selectively.

if no, then this is fine.

**Comment by paulstelian97** [comment]
@plbossart When traces are disabled they are compiled out even without this pull request. This PR only serves to allow more granularity in which of the verbose (analog to DEBUG prints on the kernel side) traces are enabled and which are not. I do see where you're coming from with the actual tracing code but the traces are supposed to be lightweight. The trace classes are not bits for easy comparison of which are enabled and which are not, and doing 34 comparisons per trace will get real expensive real fast (enough so it is no longer worth it to have this hack of a trace system even -- might as well have the messages themselves)

**Comment by plbossart** [comment]
> @plbossart When traces are disabled they are compiled out even without this pull request. This PR only serves to allow more granularity in which of the verbose (analog to DEBUG prints on the kernel side) traces are enabled and which are not. I do see where you're coming from with the actual tracing code but the traces are supposed to be lightweight. The trace classes are not bits for easy comparison of which are enabled and which are not, and doing 34 comparisons per trace will get real expensive real fast (enough so it is no longer worth it to have this hack of a trace system even -- might as well have the messages themselves)

well, maybe 34 types of traces isn't very smart in the first place...

when you need to support a product where e.g. the firmware is released and signed or part of a larger package, it's a logistical nightmare to re-distribute a new firmware just so that the developers can request more or fewer traces in bug reports. You really want to use the released firmware and still gather as much information as possible.

If the traces are only intended for developers that can recompile firmware at will, then we'd be writing off an important source of information to help with remote debug. I personally spend more time for the latter category these days.

**Comment by paulstelian97** [comment]
> > @plbossart When traces are disabled they are compiled out even without this pull request. This PR only serves to allow more granularity in which of the verbose (analog to DEBUG prints on the kernel side) traces are enabled and which are not. I do see where you're coming from with the actual tracing code but the traces are supposed to be lightweight. The trace classes are not bits for easy comparison of which are enabled and which are not, and doing 34 comparisons per trace will get real expensive real fast (enough so it is no longer worth it to have this hack of a trace system even -- might as well have the messages themselves)
> 
> well, maybe 34 types of traces isn't very smart in the first place...
> 
> when you need to support a product where e.g. the firmware is released and signed or part of a larger package, it's a logistical nightmare to re-distribute a new firmware just so that the developers can request more or fewer traces in bug reports. You really want to use the released firmware and still gather as much information as possible.
> 
> If the traces are only intended for developers that can recompile firmware at will, then we'd be writing off an important source of information to help with remote debug. I personally spend more time for the latter category these days.

I see your point. Then guess I'm asking @lgirdwood for feedback on whether I should pursue
making a runtime trace enabling/disabling (or letting @akloniex change this PR in that direction).

Although I'd argue that there are some extremely verbose traces that may still be guarded by a
compile flag (the "WFE", "WFX", and other short few-letter words ones come to mind) or even be
updated completely. Those seem to be relics from before the current trace system was even born.
What do all of you think about this situation?

**Comment by plbossart** [comment]
> > > @plbossart When traces are disabled they are compiled out even without this pull request. This PR only serves to allow more granularity in which of the verbose (analog to DEBUG prints on the kernel side) traces are enabled and which are not. I do see where you're coming from with the actual tracing code but the traces are supposed to be lightweight. The trace classes are not bits for easy comparison of which are enabled and which are not, and doing 34 comparisons per trace will get real expensive real fast (enough so it is no longer worth it to have this hack of a trace system even -- might as well have the messages themselves)
> > 
> > 
> > well, maybe 34 types of traces isn't very smart in the first place...
> > when you need to support a product where e.g. the firmware is released and signed or part of a larger package, it's a logistical nightmare to re-distribute a new firmware just so that the developers can request more or fewer traces in bug reports. You really want to use the released firmware and still gather as much information as possible.
> > If the traces are only intended for developers that can recompile firmware at will, then we'd be writing off an important source of information to help with remote debug. I personally spend more time for the latter category these days.
> 
> I see your point. Then guess I'm asking @lgirdwood for feedback on whether I should pursue
> making a runtime trace enabling/disabling (or letting @akloniex change this PR in that direction).
> 
> Although I'd argue that there are some extremely verbose traces that may still be guarded by a
> compile flag (the "WFE", "WFX", and other short few-letter words ones come to mind) or even be
> updated completely. Those seem to be relics from before the current trace system was even born.
> What do all of you think about this situation?

@mmaka1 can you chime in as well? I vaguely recall we discussed this notion of 'dynamic debug' some 18-months ago.

I am not going to lay on the tracks on this specific PR, but that's an important topic.

Note that btw I have the same concern on the kernel side. we tend to enable way too many logs, it's either errors only or a ton of stuff that is not really important. It'd like finer granularity as done for DRM/i915 where you can set some bits on which logs you care about.

File: src/audio/asrc/asrc.c (hunk 7 above)
Comment (lyakh) [reply src/audio/asrc/asrc.c]: @paulstelian97 I'd be fine keeping those defines in respective .c as long as they're used only there. I don't see a need for a common header if all or at least most of them are only used in a single file. However, if sufficiently many of them are or can be used in multiple files, then yes, we could use a common header.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
I think we've all had some discussion on this topic lately, the key points I can see is granularity (improve SNR) and runtime configuration (no need for debuggers to resign code).  I think we have agreement on several features

1) Trace should be runtime configurable
2) Trace should be configurable per component / subsystem.
3) Trace should have several levels like Linux printk(LEVEL),

This would allow us to do things like

Enable all trace upto level 4 in MUX1.1 component
```
echo "MUX1.1 level=4" > /sysfs/kernel/debug/sof/trace_events
```
or enable all trace events for debug features 1,3 and 8 in KWD 1.0
```
echo "KWD1.0 debug=1,3,8" > /sysfs/kernel/debug/sof/trace_events"
```

**Comment by jajanusz** [comment]
@akloniex I don't know if it still makes sense. It may need some update IRT @mmaka1 work with classes/UUID.

**Comment by lgirdwood** [comment]
> @akloniex I don't know if it still makes sense. It may need some update IRT @mmaka1 work with classes/UUID.

@mmaka1 can you give us an example usage using the classes/UUID. I dont really want user to have to map a topology name to a UUID and vice versa in order to use the tool..
